### PR TITLE
Make ASTWriter, BinaryWriter, BinaryWriterSpec C++ classes

### PR DESCRIPTION
**Pull request by binji**

### CODE REVIEW

File: src/ast-writer.cc [hunk 1]
```diff
@@ -58,302 +58,338 @@ enum class NextChar {
...
+  assert(indent_ >= 0);
 }
 
-static void write_indent(Context* ctx) {
+void ASTWriter::WriteIndent() {
   static char s_indent[] =
       "                                                                       "
       "                                                                       ";
   static size_t s_indent_len = sizeof(s_indent) - 1;
-  size_t indent = ctx->indent;
```
Comment (KarlSchimpf) [inline src/ast-writer.cc]: Why did you move from a local variable to the field? Doesn't this change the effect (you change the indent of the writer)?

Comment (binji) [inline src/ast-writer.cc]: Hm, good point. Wonder why this wasn't breaking tests?
